src/http/local_async_normal.hpp

```diff
@@ -254,177 +267,213 @@ struct http_async_connection<Tag, version_major, version_minor>
+            // reached while reading the socket. This signals the end of the
+            // body processing chain.
             if (callback) {
               typename protocol_base::buffer_type::const_iterator
                   begin = this->part.begin(),
                   end = begin;
               std::advance(end, bytes_transferred);
 
-              // We call the callback function synchronously passing the
-              // error
-              // condition (in this case, end of file) so that it can
-              // handle
-              // it appropriately.
-              callback(make_iterator_range(begin, end), ec);
+              // We call the callback function synchronously passing the error
+              // condition (in this case, end of file) so that it can handle it
+              // appropriately.
+              if (this->is_chunk_encoding && remove_chunk_markers_) {
+                callback(parse_chunk_encoding(make_iterator_range(begin, end)), ec);
+              } else {
+                callback(make_iterator_range(begin, end), ec);
+              }
             } else {
               string_type body_string;
-              std::swap(body_string, this->partial_parsed);
-              body_string.append(this->part.begin(), bytes_transferred);
-              if (this->is_chunk_encoding)
-                this->body_promise.set_value(parse_chunk_encoding(body_string));
-              else
+              if (this->is_chunk_encoding && remove_chunk_markers_) {
+                for (size_t i = 0; i < this->partial_parsed.size(); i += 1024) {
+                  auto range = parse_chunk_encoding(boost::make_iterator_range(
+                      static_cast<typename std::array<typename char_<Tag>::type, 1024>::const_iterator>(
+                          this->partial_parsed.data()) + i,
+                      static_cast<typename std::array<typename char_<Tag>::type, 1024>::const_iterator>(
+                          this->partial_parsed.data()) +
+                          std::min(i + 1024, this->partial_parsed.size())));
+                  body_string.append(boost::begin(range), boost::end(range));
+                }
+                this->partial_parsed.clear();
+                auto range = parse_chunk_encoding(boost::make_iterator_range(
+                    this->part.begin(),
+                    this->part.begin() + bytes_transferred));
+                body_string.append(boost::begin(range), boost::end(range));
+                this->body_promise.set_value(body_string);
+              } else {
+                std::swap(body_string, this->partial_parsed);
+                body_string.append(this->part.begin(),
+                                   this->part.begin() + bytes_transferred);
                 this->body_promise.set_value(body_string);
+              }
             }
-            // TODO set the destination value somewhere!
+            // TODO(dberris): set the destination value somewhere!
```

Comment:
Is dberris actually going to help here? If not, can we assign the todo to someone who will?

Reply:
And here.